Bulma modal prompts

Geo-Bold opened the pull request:
Resolves the use of Bulma modals instead of JS prompts while retaining as much initial structure as possible. Modal creating function is easily reusable and is not rendered to the DOM until called. A garbage collection function ensures that no duplicated modals remain rendered at one time.

For the local storage reset protocol, there is similar structure to the original JS prompts and confirms. So far, there exists control over the prompt and establishment of a generic default.

Further changes would be centered around enhancing the functionality of the modal creating function to allow for more control over each modal generated as well as revisiting the current style for the prompts to include two buttons, making them larger and removing the input field when not necessary.

Geo-Bold committed:
feat: adds generic bulma modal functions to create user prompts and append them to  the dom.
feat: adds bulma modal prompt for user incident name and saves to local storage.
feat: adds bulma modal prompts for clearing local storage.
style
feat: provides a method for disposing of existing modal prompts from the dom.

Geo-Bold committed:
feat: renames inputUserIncident to modalPrompt and makes it more clean and concise.
feat: renames resetProtocol to confirmReset and revamps to use a single generic function. promises allow the calling function to await user input.
issues: needs broader controls on the modal creation functions and two button controls for positive and negative states

anthonyisensee commented:
@Geo-Bold thanks so much for this! Thanks for the detailed description, as well. I'll review it and add some feedback as soon as possible!

Geo-Bold committed:
Merge branch 'bulma-modal-prompts' into prod

Geo-Bold commented:
The new commit resolves the conflict from @anthonyisensee  prod branch.

Geo-Bold committed:
feat: sets the trigger for incident prompt back to the set service name.
refactor: calling the createModal function now creates a visible prompt by default

Geo-Bold committed:
Total Modal Card Overhaul
feat: adds a default modal with improved style and two button functionality. includes option to obtain user input.
style: improves user prompt messages.
refactor: improves modal.js source code readability.

anthonyisensee commented:
Thanks for the pull request, @Geo-Bold! I can tell you put a lot of time into it, and I really appreciate it. Thank you, as well, for proactively resolving the merge conflict, as well as for your Total Modal Card Overhaul commit. That commit added a number of things I was going to request, mainly the (nearly) complete card modal with a separate modal layout for prompt and confirm!

A professor of mine once gave me some key insight that has stuck with me in all my programming since. After programming, or when reviewing others code, you have three main questions to answer:

* Does it work/run/meet the requirements? If yes, 50% of the potential value has been achieved.
* Is it readable and maintainable? - If yes, 40% of the potential value has been achieved.
* Is it well optimized? - If yes, 10% of the potential value has been achieved.

These insights guide my feedback.

### Does it work?

Yes! It does, and that warrants a massive kudos to you and your work. I love how responsive the modal feels, and to the end user, that's all that matters. You've delivered the client's ask, and for that they love you! With this alone, you've achieved 50% of the goal. It is important to note this success! Anything that follows this is (in my mind, mandatory) extra credit that ensures the code is maintainable and readable.

### Is it readable and maintainable?

Sort of. 

**First of all**, the thing I love. You recognized that modal functionality was different enough from the `Main.js` or any other piece of the project that it warranted it's own file. I wasn't necessarily thinking of doing this, but it was totally the right decision!

That said, there are definitely some improvements that can be made!

To understand this, the idea of [software orthogonality](https://en.wikipedia.org/wiki/Orthogonality_(programming)) is crucial. The key idea of software orthogonality is that individuals pieces of the software can grow all on their own (or even be entirely replaced) without impacting the rest of the software. In short, that their purposes are orthogonal.

Consider this illustration of two orthogonal vectors. Because they are separated, they can grow and scale in functionality all on their own without impacting the other, or one requiring modification of the other. Additionally, because they are scaling separately, they become deeply reusable throughout the application for purposes that aren't specific to just one file or function, usable throughout the codebase without impacting other pieces of the code. And, because they aren't tied to each other, each can grow in its own direction at the fastest pace possible.

Ideal:

![image](https://github.com/anthonyisensee/incidenttracker/assets/60198052/f5186129-a27d-4d45-84de-c0eb40b06372)

At the moment, however, the purposes of the two files are tied together. If one is modified, so must the other be. Code specific to `Main.js` is being run inside of `Modal.js` ([here](https://github.com/Geo-Bold/incidenttracker/blob/bulma-modal-prompts/Modal.js#L151-L157)), meaning that in order to modify the functionality of `Main.js` one must edit `Modal.js`. As such, the vectors are no longer orthogonal, but are tied together. This means that they are no longer separately scalable, and that changes to one deeply impacts the other. This is not ideal! 

Current state:

![image](https://github.com/anthonyisensee/incidenttracker/assets/60198052/f9d6193c-b40f-4a10-957a-fd87b8a7943d)

I think that you intuitively knew this. My evidence is that you used [promises](https://github.com/Geo-Bold/incidenttracker/blob/bulma-modal-prompts/Modal.js#L107) and [their callbacks](https://github.com/Geo-Bold/incidenttracker/blob/bulma-modal-prompts/Main.js#L177-L202) for the confirmation modal. While I would like to see the promises removed (as adding [asynchronous programming](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Introducing) to such a simple, locally running application with so little blocking operations needlessly increases the complexity) I love that you wanted to keep `Main.js` code inside of `Main.js`.

**Secondly**, let's talk object oriented (for a OO JS guide, I highly recommend reading through [this example](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/Object-oriented_programming) on mdn). You decided to export functions from `Modal.js` and import them into a Modal object in `Main.js`. While this does work, another way is to make `Modal.js` a full on separate class in its respective file and add a constructor. (For a rough template, see [this code](https://github.com/anthonyisensee/moneytimer/blob/main/js/TimeTracker.js). Just note, since writing that code I've discovered that prefixing a variable with `_` is no longer recommended, so you don't need to do that).

In the following, key callouts should be made to **Objects**, **Methods**, and **Properties**. These are three of the biggest components of proper Object Oriented programming.

```js
// Modal.js defines a class that initializes and provides functionality to Modal objects
export class Modal() {

    // Runs once upon object instantiation/initialization. Useful for setting properties associated with the object as a whole.
    constructor(...){
        
        // Dynamically set a property of the class
        this.modal = document.getElementById("modal")

        ...

    }

    // Properties can also be set outside of the constructor, but you do so without using variable declaration (const, let, or var)
    example_property = "example"

    // Class methods are the things a class can do. Note that they do not require the "function" keyword
    prompt(...) {
        ...
    }
    
    confirm(...) {
        ...
    }
    
}
```

This functionality can then be be imported and used in other classes like so:

```js
import { Modal } from "example/path/Modal.js";

...

const modal = new Modal()

```

I would request that since the rest of the repo is using OO, that you attempt to implement it with `Modal.js` as well. If you have any questions, definitely do some research into OO and/or feel free to ask questions. The sooner you get exposure to proper OO, the better! Just keep in mind as you build the class that it could be instantiated at the same time from other code multiple times, and that you should treat the modal state (open or closed) as an unknown.

**Thirdly**, there is a way to make [this portion of your code](https://github.com/Geo-Bold/incidenttracker/blob/bulma-modal-prompts/Modal.js#L4-L77) significantly shorter, more readable, and modifiable. It will also replace your documentation with the code itself, which is always a massive win!

Instead of dividing the functionality up by individual elements:

```js
function createExampleDiv(message) {

  const div = document.createElement('div')  // Layout
  div.classList.add("a", "b", "c")  // Layout
  div.setAttribute("id", "example_div")  // Layout
  div.addEventListener(...)  // Functionality

  const p = document.createElement('p')  // Layout
  p.classList.add('e')  // Layout
  p.innerHTML = message  // Layout
  p.addEventListener(...)  // Functionality
  
  div.appendChild(p)  // Layout
  
  return div

}
```

Divide the functionality up directly:

```js

function createExampleDivBetter(message) {

  // Layout
  const html = `
    <div id="example_div" class="a b c">
      <p class="e">${message}</p>
    </div>
  `
  const node = new DOMParser().parseFromString(html, 'text/html').body.firstChild

  // Functionality
  node.getElementById("example_div").addEventListener(...)
  node.querySelector("p").addEventListener(...)

  return node

}

```

This will _significantly_ simplify your layout life, especially after you get 20 layers of HTML deep and/or have come back to the code after you've forgotten exactly what the HTML layout looks like. For an example, see [this code](https://github.com/anthonyisensee/moneytimer/blob/dev/js/TimesheetPanel.js#L76-L161). Just think how miserable it would be trying to keep this all in your head and modify it in the format of the first example!

**Finally**, a minor tweak. To maintain consistency all code should be four space indented. Wars have been waged over 2 vs 4 space indenting, and both have advantages and disadvantages. But, as much as possible, you should do your best to maintain the style of a repository you are contributing to. If you want to make your life easier, you can set workspace specific indenting rules in VSCode (or most other IDEs) so it will handle everything for you automatically in only your one workspace.

There are other components that I may circle back around to depending on your implementation of OO, but for now these are the big things.

### Is it well optimized?

It's JS. As long as there aren't any glaring optimization issues, I and most of the rest of the world could care less. ;D

### Additional Specifications

I'm going to do something that most normal clients never do; give you specific coding asks for a product. My hope is that they will be helpful! :)

1. Make your code OO, with clear orthogonality between `Main.js` and `Modal.js`. Once your done, I want `Modal.js` to be usable in any of your future (bulma) projects by adding the `Modal.js` file alone.
2. Add `prompt()` and `confirm()` methods to the Modal class (please avoid `modalPrompt()` and `modalConfirm()` as such naming schemas are redundant and force the developer to type modal over and over, such as with `modal.modalPrompt()`).
3. Add a parameter to `confirm()` and `prompt()` for a) a generic function that runs upon success and b) a generic function that runs upon failure. These functions should be executed when the user clicks the corresponding button in the modal, and the generic functions for prompt should have the results of the input field passed in to them before execution. I believe this will be key in increasing orthogonality and will allow you to remove Promises from `Modal.js` while keeping `Main.js` functionality in `Main.js`. This _is_ a bit of a complicated ask, so definitely reach out if you have questions.
4. Implement the layout/functionality separation I suggested.
5. Add the modal card footer to the modal and move any buttons to it. For prompt and confirm, please add the `is-danger` class to the confirm button and the `is-info` class to the cancel button.

Thanks again! Like I said in "Does it work?" you've done great so far. The solution you've delivered is working and end users that got a preview are happy with the feature! That's 50% of the battle, right there. That said, before the code is ready for merge and to be rolled out to all users, addressing that last 40% is important!

If you have any questions, feel free to drop a comment here or reach out via separate channels!

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,199 @@
+
+  
+  // Add a click event on various child elements to delete the parent modal
+  const altCloseModal = (document.querySelectorAll('.modal-background, .modal-close, .modal-card-head .delete, .modal-card-foot, .exit') || []).forEach(($close) => {
```
@Geo-Bold is there a particular reason you're using `|| []` here? If the query selector doesn't match anything it will return an empty array by default.

Geo-Bold replied:
@anthonyisensee No reason. I imported as is from Bulma's modal function library.

anthonyisensee replied:
I think all instances of `|| []` can be removed, then.

Tested to confirm with `document.querySelectorAll('#bogus_id').forEach(e => { console.log(e) })` run in the browser console. the `.forEach(...)` simply doesn't log anything.

Geo-Bold commented:
> Thanks for the pull request, @Geo-Bold! I can tell you put a lot of time into it, and I really appreciate it. Thank you, as well, for proactively resolving the merge conflict, as well as for your Total Modal Card Overhaul commit. That commit added a number of things I was going to request, mainly the (nearly) complete card modal with a separate modal layout for prompt and confirm!
> 
> A professor of mine once gave me some key insight that has stuck with me in all my programming since. After programming, or when reviewing others code, you have three main questions to answer:
> 
> * Does it work/run/meet the requirements? If yes, 50% of the potential value has been achieved.
> * Is it readable and maintainable? - If yes, 40% of the potential value has been achieved.
> * Is it well optimized? - If yes, 10% of the potential value has been achieved.
> 
> These insights guide my feedback.
> 
> ### Does it work?
> Yes! It does, and that warrants a massive kudos to you and your work. I love how responsive the modal feels, and to the end user, that's all that matters. You've delivered the client's ask, and for that they love you! With this alone, you've achieved 50% of the goal. It is important to note this success! Anything that follows this is (in my mind, mandatory) extra credit that ensures the code is maintainable and readable.
> 
> ### Is it readable and maintainable?
> Sort of.
> 
> **First of all**, the thing I love. You recognized that modal functionality was different enough from the `Main.js` or any other piece of the project that it warranted it's own file. I wasn't necessarily thinking of doing this, but it was totally the right decision!
> 
> That said, there are definitely some improvements that can be made!
> 
> To understand this, the idea of [software orthogonality](https://en.wikipedia.org/wiki/Orthogonality_(programming)) is crucial. The key idea of software orthogonality is that individuals pieces of the software can grow all on their own (or even be entirely replaced) without impacting the rest of the software. In short, that their purposes are orthogonal.
> 
> Consider this illustration of two orthogonal vectors. Because they are separated, they can grow and scale in functionality all on their own without impacting the other, or one requiring modification of the other. Additionally, because they are scaling separately, they become deeply reusable throughout the application for purposes that aren't specific to just one file or function, usable throughout the codebase without impacting other pieces of the code. And, because they aren't tied to each other, each can grow in its own direction at the fastest pace possible.
> 
> Ideal:
> 
> ![image](https://private-user-images.githubusercontent.com/60198052/320809882-f5186129-a27d-4d45-84de-c0eb40b06372.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI2NzQ4MjksIm5iZiI6MTcxMjY3NDUyOSwicGF0aCI6Ii82MDE5ODA1Mi8zMjA4MDk4ODItZjUxODYxMjktYTI3ZC00ZDQ1LTg0ZGUtYzBlYjQwYjA2MzcyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA5VDE0NTUyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdlMTYyYzFmNDE5YTRhM2QyZDQ4NWIzZmQ3MzRiNWY2M2Y5YzAwZmU3YzhjMGMyYzRmODU4ODdlODU5NWQyMGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.COggy1l3BMVbJCLzsl2VGYdFXAOXy_PuZ5rpwa7t2g0)
> 
> At the moment, however, the purposes of the two files are tied together. If one is modified, so must the other be. Code specific to `Main.js` is being run inside of `Modal.js` ([here](https://github.com/Geo-Bold/incidenttracker/blob/bulma-modal-prompts/Modal.js#L151-L157)), meaning that in order to modify the functionality of `Main.js` one must edit `Modal.js`. As such, the vectors are no longer orthogonal, but are tied together. This means that they are no longer separately scalable, and that changes to one deeply impacts the other. This is not ideal!
> 
> Current state:
> 
> ![image](https://private-user-images.githubusercontent.com/60198052/320841290-f9d6193c-b40f-4a10-957a-fd87b8a7943d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI2NzQ4MjksIm5iZiI6MTcxMjY3NDUyOSwicGF0aCI6Ii82MDE5ODA1Mi8zMjA4NDEyOTAtZjlkNjE5M2MtYjQwZi00YTEwLTk1N2EtZmQ4N2I4YTc5NDNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA5VDE0NTUyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTczNTI1NDllOTc5MjQ0NDQzMjcxMDRmOWVjMGM0MzRjZmNhNmJlZjhlZDIwZGYzZmRmNTJiNjRiYjBjODY4ZTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.E22aoY1lY2It9gHCqrUoX712Io7khSBLGek2AHRx9y4)
> 
> I think that you intuitively knew this. My evidence is that you used [promises](https://github.com/Geo-Bold/incidenttracker/blob/bulma-modal-prompts/Modal.js#L107) and [their callbacks](https://github.com/Geo-Bold/incidenttracker/blob/bulma-modal-prompts/Main.js#L177-L202) for the confirmation modal. While I would like to see the promises removed (as adding [asynchronous programming](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Introducing) to such a simple, locally running application with so little blocking operations needlessly increases the complexity) I love that you wanted to keep `Main.js` code inside of `Main.js`.
> 
> **Secondly**, let's talk object oriented (for a OO JS guide, I highly recommend reading through [this example](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/Object-oriented_programming) on mdn). You decided to export functions from `Modal.js` and import them into a Modal object in `Main.js`. While this does work, another way is to make `Modal.js` a full on separate class in its respective file and add a constructor. (For a rough template, see [this code](https://github.com/anthonyisensee/moneytimer/blob/main/js/TimeTracker.js). Just note, since writing that code I've discovered that prefixing a variable with `_` is no longer recommended, so you don't need to do that).
> 
> In the following, key callouts should be made to **Objects**, **Methods**, and **Properties**. These are three of the biggest components of proper Object Oriented programming.
> 
> ```js
> // Modal.js defines a class that initializes and provides functionality to Modal objects
> export class Modal() {
> 
>     // Runs once upon object instantiation/initialization. Useful for setting properties associated with the object as a whole.
>     constructor(...){
>         
>         // Dynamically set a property of the class
>         this.modal = document.getElementById("modal")
> 
>         ...
> 
>     }
> 
>     // Properties can also be set outside of the constructor, but you do so without using variable declaration (const, let, or var)
>     example_property = "example"
> 
>     // Class methods are the things a class can do. Note that they do not require the "function" keyword
>     prompt(...) {
>         ...
>     }
>     
>     confirm(...) {
>         ...
>     }
>     
> }
> ```
> 
> This functionality can then be be imported and used in other classes like so:
> 
> ```js
> import { Modal } from "example/path/Modal.js";
> 
> ...
> 
> const modal = new Modal()
> ```
> 
> I would request that since the rest of the repo is using OO, that you attempt to implement it with `Modal.js` as well. If you have any questions, definitely do some research into OO and/or feel free to ask questions. The sooner you get exposure to proper OO, the better! Just keep in mind as you build the class that it could be instantiated at the same time from other code multiple times, and that you should treat the modal state (open or closed) as an unknown.
> 
> **Thirdly**, there is a way to make [this portion of your code](https://github.com/Geo-Bold/incidenttracker/blob/bulma-modal-prompts/Modal.js#L4-L77) significantly shorter, more readable, and modifiable. It will also replace your documentation with the code itself, which is always a massive win!
> 
> Instead of dividing the functionality up by individual elements:
> 
> ```js
> function createExampleDiv(message) {
> 
>   const div = document.createElement('div')  // Layout
>   div.classList.add("a", "b", "c")  // Layout
>   div.setAttribute("id", "example_div")  // Layout
>   div.addEventListener(...)  // Functionality
> 
>   const p = document.createElement('p')  // Layout
>   p.classList.add('e')  // Layout
>   p.innerHTML = message  // Layout
>   p.addEventListener(...)  // Functionality
>   
>   div.appendChild(p)  // Layout
>   
>   return div
> 
> }
> ```
> 
> Divide the functionality up directly:
> 
> ```js
> function createExampleDivBetter(message) {
> 
>   // Layout
>   const html = `
>     <div id="example_div" class="a b c">
>       <p class="e">${message}</p>
>     </div>
>   `
>   const node = new DOMParser().parseFromString(html, 'text/html').body.firstChild
> 
>   // Functionality
>   node.getElementById("example_div").addEventListener(...)
>   node.querySelector("p").addEventListener(...)
> 
>   return node
> 
> }
> ```
> 
> This will _significantly_ simplify your layout life, especially after you get 20 layers of HTML deep and/or have come back to the code after you've forgotten exactly what the HTML layout looks like. For an example, see [this code](https://github.com/anthonyisensee/moneytimer/blob/dev/js/TimesheetPanel.js#L76-L161). Just think how miserable it would be trying to keep this all in your head and modify it in the format of the first example!
> 
> **Finally**, a minor tweak. To maintain consistency all code should be four space indented. Wars have been waged over 2 vs 4 space indenting, and both have advantages and disadvantages. But, as much as possible, you should do your best to maintain the style of a repository you are contributing to. If you want to make your life easier, you can set workspace specific indenting rules in VSCode (or most other IDEs) so it will handle everything for you automatically in only your one workspace.
> 
> There are other components that I may circle back around to depending on your implementation of OO, but for now these are the big things.
> 
> ### Is it well optimized?
> It's JS. As long as there aren't any glaring optimization issues, I and most of the rest of the world could care less. ;D
> 
> ### Additional Specifications
> I'm going to do something that most normal clients never do; give you specific coding asks for a product. My hope is that they will be helpful! :)
> 
> 1. Make your code OO, with clear orthogonality between `Main.js` and `Modal.js`. Once your done, I want `Modal.js` to be usable in any of your future (bulma) projects by adding the `Modal.js` file alone.
> 2. Add `prompt()` and `confirm()` methods to the Modal class (please avoid `modalPrompt()` and `modalConfirm()` as such naming schemas are redundant and force the developer to type modal over and over, such as with `modal.modalPrompt()`).
> 3. Add a parameter to `confirm()` and `prompt()` for a) a generic function that runs upon success and b) a generic function that runs upon failure. These functions should be executed when the user clicks the corresponding button in the modal, and the generic functions for prompt should have the results of the input field passed in to them before execution. I believe this will be key in increasing orthogonality and will allow you to remove Promises from `Modal.js` while keeping `Main.js` functionality in `Main.js`. This _is_ a bit of a complicated ask, so definitely reach out if you have questions.
> 4. Implement the layout/functionality separation I suggested.
> 5. Add the modal card footer to the modal and move any buttons to it. For prompt and confirm, please add the `is-danger` class to the confirm button and the `is-info` class to the cancel button.
> 
> Thanks again! Like I said in "Does it work?" you've done great so far. The solution you've delivered is working and end users that got a preview are happy with the feature! That's 50% of the battle, right there. That said, before the code is ready for merge and to be rolled out to all users, addressing that last 40% is important!
> 
> If you have any questions, feel free to drop a comment here or reach out via separate channels!

Geo-Bold committed:
Modal Card Optimization Patch
refactor: removes dependencies and rewrites the modal source code to maximize orthogonality to main.
feat: adds the modal class along with prompt and confirm methods.
feat: includes method arguments to execute generic function using callbacks.
refactor: modal card is now parsed as an html string, improving structure and code readability.
feat: adds the 'is-danger' and 'is-info' classes to the confirm and cancel buttons.

Geo-Bold commented:
@anthonyisensee Thank you for your thorough feedback. It was instrumental in directing my efforts for this patch. I am sure the client will be more than satisfied :)

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,172 @@
+export class Modal {
+
+    modal = `
```
This is so much more readable for layout! I love it.

Geo-Bold replied:
Ugly? Yes. Readable? Absolutely!

anthonyisensee replied:
You'll see the light someday. ;P

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,172 @@
+        this.secondaryBtnText = secondaryBtnText || 'Cancel'
+    }
+
+    #createTemplate() {
```
Nice!! You learned about private methods. This is a _fantastic_ use of a private method

This is also a fantastic example of writing Don't Repeat Yourself (or DRY) code. Great job recognizing that functionality for prompt and confirm could be abstracted out a layer! Choices like this reap great rewards after a codebase has grown significantly.

Geo-Bold replied:
Thank you! I will be sending this to my professor. 😛

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,172 @@
+                    <p class="modal-card-title"></p>
+                </header>
+                <section class="modal-card-body py-5">
+                    <p class="modal-input is-size-5 mb-1"></p>
+                    <input id="modal-input" class="input">
```
I would recommend clearing out this content and setting it dynamically in the prompt and confirm methods! Look into using [Bulma form controls](https://bulma.io/documentation/form/general/) for the input field (especially using `div.field` and a contained `.label`) and [Bulma content container](https://bulma.io/documentation/elements/content/) for normal text.

Geo-Bold replied:
I like it! I'll be appending those capabilities to the current build.

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,172 @@
+                    <input id="modal-input" class="input">
+                </section>
+                <footer class="modal-card-foot py-3">
+                    <div class="modal-controls is-flex is-justify-content-end is-flex-grow-1">
```
Look into using the Bulma `div.buttons` container here.

Geo-Bold replied:
I changed the modal structure to use more of Bulma's built in functionality by including field and control elements to automatically space the modal contents.

anthonyisensee replied:
Nice! Looking forward to seeing it!

anthonyisensee commented on Main.js:
```diff
@@ -165,34 +167,45 @@ document.addEventListener('DOMContentLoaded', () => {
-            if (confirm("Are you absolutely sure you want to reset everything? This can NOT be undone.")) {
+            e && confirmReset_2.confirm((e) => {
 
-                if (prompt(`Type "Anthony is the best" to confirm that you definitely, absolutely, certainly wish to reset everything.`) == "Anthony is the best") {
```
Sad to see it go, but since it's no longer _just_ my project, I can't hide such obvious easter eggs in it.

Geo-Bold replied:
Haha, sorry. On the bright side, the user can set up whatever prompt response they wish.

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,172 @@
+        const modalMessage = parsedContent.querySelector('.modal-input')
+        modalMessage.innerText = this.message
+
+        if (!(this.requireUserInput)) {
```
Evaluation happens from right to left. Properties are evaluated with their parent object. This makes the inner `()` unnecessary. Imo the `()` also makes the statement less readable, but if it makes it more readable to you then feel free to keep it. Just a consideration.

Geo-Bold replied:
The source code has been changed to check for a positive state as opposed to the negative. Also, the input is now parsed and appended inside this selection node as opposed to automatically being appended as part of the default modal and removed if not required.

anthonyisensee replied:
You found a way to make it even more readable! Nice!!

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,172 @@
+        document.body.appendChild(modalDiv)
+
+        // Event handling to close the modal with the mouse.
+        document.querySelectorAll('.modal-background, .modal-close, .modal-card-head .delete, .modal-card-foot, .exit').forEach(($close) => {
+
+            const $target = $close.closest('.modal')
+
+            $close.addEventListener('click', () => { deleteModal($target) })
+
+        }) 
```
The `$` is an allowed character in variable names, but it is [commonly used to indicate variables that are JQuery objects](https://stackoverflow.com/a/553734/13173288) (gross). Let's not make people think we're using JQuery. Also, when copy/pasta-ing code, make sure to review the code entirely and make sure it meets your standards.

Geo-Bold replied:
Variable names have been amended.

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,172 @@
+
+    }
+
+    prompt(userFunction = undefined, alternateFunction = undefined) {
```
Remember that another developer is going to be implementing this code, preferably without having to read through it. Since this is a public method, we want its parameters to make the most sense in a public context for use from outside the scope of this class. 

If I was a developer implementing your code, userFunction and alternateFunction would give me no clue as to what the functions should do or what they were executing. Instead, I'd want these function parameters to be `functionToRunOnOkay` and `functionToRunOnCancel`. A bit long, yes, but they leave no doubt in terms of what these parameters should be and when they will run.

If you want to go the extra mile, you can add [jsdocs styled function documentation](https://jsdoc.app/about-getting-started) (see the "Adding more information with tags" section). This would allow you to explain the parameters, when they get called, and importantly, that the user input will be provided as the first function parameter by default.

Geo-Bold replied:
I will admit, I do shy away from verbose function or variable names, but I do understand your point. I will create less ambiguous and more descriptive names in the future.

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,172 @@
+
+            if (userFunction) {
+
+                return userFunction(userInput.value)
```
Fantastic job figuring out how to use abstracted functions. I'm impressed that you figured out how to pass that input value into it and then return the function so that a callback function can handle the input. Great job!

Take some time to consider if instead of returning the function, executing the function right here would be best. I'm not sure that it would be for prompt, because this has the added benefit of allowing you to process the input in the callback in `Main.js`, but it's something to consider for confirm. At the very least, know that you _could_ execute the function here instead of returning it!

Geo-Bold replied:
Your appreciation is appreciated, greatly!

anthonyisensee replied:
Chat-GPT, that you? 😜

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,172 @@
+        </div> 
+    `
+
+    constructor(message, title, input, primaryBtnText, secondaryBtnText) {
```
I'd move the `input` param to the end here, as it is the least likely to get used.

Geo-Bold replied:
The input parameter has been removed. The modal defaults with no input included. Instead, the input element is created and appended to the modal only when the prompt method is called.

anthonyisensee replied:
Even better!

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,172 @@
+
+    constructor(message, title, input, primaryBtnText, secondaryBtnText) {
+
+        this.title = title || 'Attention' 
+        this.message = message || 'This is the default alert message. Have a great day!'
+        this.requireUserInput = input || false
+        this.primaryBtnText = primaryBtnText || 'OK'
+        this.secondaryBtnText = secondaryBtnText || 'Cancel'
```
Read [this](https://stackoverflow.com/a/61481631/13173288) on `||` vs `??`. When setting defaults in the body, I'd shy away from `||` and instead try to use `??` as it's results are more consistent and expected.

Also, a useful JS thing to know about are [objects built to be parameters](https://stackoverflow.com/a/7764592/13173288). I've seen libraries whose functions/methods don't accept straight up parameters at all, and instead only accept a parameter object. While this may at first seem like a needless complication, it becomes massively powerful in that _parameter order no longer matters_. Also, it makes setting defaults wicked easy as you simply check for the property in the object and if it doesn't exist, use a default (i.e. in a function `func(paramObject){...}` using something like `const content = paramObject.content ?? "<p>Content</p>"`).

Geo-Bold replied:
I have change the default operators in favor of the more appropriate _nullish co-ops_.
Our conversation about using objects as parameters was mind-blowing. I will be using that as my primary parameter style when requiring more than three arguments.

anthonyisensee replied:
cough cough _nullish coalescing_

Yeah, parameter objects are OP!! Here's the [project I referenced](https://slimselectjs.com/settings#select) if you want to see how somebody else does it.

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,172 @@
+        
+        primaryButton.addEventListener('click', () => {
+
+            if (userFunction) {
```
Good check! Moving this if outside of the event listener will prevent creating an event listener that does nothing, however. Same with other places you do this.

Geo-Bold replied:
You're wicked! Nice logic catch. Changes have been made.

anthonyisensee replied:
😈

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,172 @@
+
+}
+
+
+
+
+  
+  
```
Unnecessary whitespace.

Geo-Bold replied:
"Hey Siri, remind me to remind Anthony to drink his coffee before reviewing any code I write. Thank you!"

anthonyisensee replied:
BuT iT sAvEs SpAcE

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,172 @@
+
+export function closeAllModals() {
+
+  (document.querySelectorAll('.modal') || []).forEach(($modal) => {
```
Unecessary `||`.

Geo-Bold replied:
Function has since been deleted. I told him he was a bad, bad boy.

anthonyisensee commented on Main.js:
```diff
@@ -165,34 +167,45 @@ document.addEventListener('DOMContentLoaded', () => {
-                    storage.remove("service_name")
-                    storage.remove("incidents")
-                    location.reload()
+                e && confirmReset_3.prompt((e) => {
```
While `value && object.method()` does indeed show that you're clever, it is far less human readable than:

```js
if (e) {
    confirmReset_2.confirm(...)
}
``` 

If readability is the only thing impacted by syntax, then increasing readability is _always_ better.

Geo-Bold replied:
Haha, I wasn't trying to be cheeky. I think that nested if statement hell may be more intuative, but is more difficult to navigate visually and personally, irritates me more.
```javascript
confirmReset_1.confirm((e) => { 
            if (e) {
                confirmReset_2.confirm((e) => {
                    if (e) {
                        confirmReset_3.prompt((e) => {
                            if (e === 'YES') {
                                storage.remove("service_name")
                                storage.remove("incidents")
                                location.reload()
                            }
                        })
                    }
                })
            }
         })
```
versus...
```javascript
confirmReset_1.confirm((e) => { 
            e && confirmReset_2.confirm((e) => {
                e && confirmReset_3.prompt((e) => {
                    e === 'YES' && (
                        storage.remove("service_name"),
                        storage.remove("incidents"),
                        location.reload()
                    )
                })
            })
        })
```

Geo-Bold replied:
https://media1.tenor.com/m/u8YEMwIfJGMAAAAC/thanos.gif

Geo-Bold replied:
![Alt Text](https://media1.tenor.com/m/u8YEMwIfJGMAAAAC/thanos.gif)

anthonyisensee replied:
I was looking at this and was thinking that both of our ways are bad.

I wrote the following method for Modal.js and will do a commit after we merge your code.

In Modal.js:

```js
/**
 * Returns a boolean result of multiple confirm modals ending with a user prompt for input.
 * @param {Array.<string>} messages - An array of strings for which to show confirmation prompts. The last string will be the input prompt.
 * @param {string} finalInputText - The text that must be matched for the final input prompt.
 * @param {function} runOnFinalConfirmation - A function that will be executed when the confirmation is totally completed.
 */
multipleConfirm(messages, finalInputText, runOnFinalConfirmation) {

    const not_last_message = messages.length != 1
    const message = messages[0]

    if (not_last_message) {  // Recursive general case

        const confirm_modal = new Modal(message)

        confirm_modal.confirm(wasConfirmed => {

            if (wasConfirmed) {

                messages.shift()

                this.multipleConfirm(messages, finalInputText, runOnFinalConfirmation)

            }

        })

    } else {                // Recursive exit case

        const prompt_modal = new Modal(message, 'Danger', true)

        prompt_modal.prompt(userInput => {

            if (userInput === finalInputText) {

                runOnFinalConfirmation()

            }

        })

    }

}
```

Main.js:

```js
// Reset protocol that creates bulma modals that confirm user wishes to clear local data
document.getElementById('reset_everything').addEventListener("click", () => {

    function deleteEverything() {

        storage.remove("service_name")
        storage.remove("incidents")
        location.reload()

    }

    const modal = new Modal.Modal()
    const messages = ['Do you wish to reset local storage?', 'Resetting will delete all data. Are you sure?', 'Please type "YES" to delete everything.']
    modal.multipleConfirm(messages, "YES", deleteEverything)

})
```


This depends on the `.prompt()` and `.confirm()` methods executing the function that is passed to it instead of returning it, so please do implement that change.

anthonyisensee replied:
At @Geo-Bold's request:

![tumblr_f456e858b7685f8ee0f6152dbc6d8f5f_2cdfda65_500](https://github.com/anthonyisensee/incidenttracker/assets/60198052/d8af3469-59a5-4baf-9feb-cad3a9020ca5)

Geo-Bold replied:
![Alt Text](https://media1.tenor.com/m/ZwUbKktJq4YAAAAC/thanos-iron-man.gif)

anthonyisensee commented on Main.js:
```diff
@@ -165,34 +167,45 @@ document.addEventListener('DOMContentLoaded', () => {
+    document.getElementById('reset_everything').addEventListener("click", () => {
 
-    document.getElementById("reset_everything").addEventListener("click", () => {
+        const confirmReset_1 = new Modal.Modal('Do you wish to reset local storage?')
+        const confirmReset_2 = new Modal.Modal('Resetting will delete all data. Are you sure?')
+        const confirmReset_3 = new Modal.Modal("Please type 'YES' to proceed.", 'Danger', true)
```
The names of these variables mix [camel and snake case](https://www.freecodecamp.org/news/programming-naming-conventions-explained/). I'd recommend using just one case. I need to get better about using one case across my whole project, but generally for local variable declarations I use snake case.

Geo-Bold replied:
Considering I've never used snake case before, it may be possible that after reading through your code, your naming convention transfered to me like some viral disease. I will be changing the names and taking a round of antibiotics!

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,172 @@
+}
+
+// Functions to open and close a modal
+export function openModal($el) {
+
+  $el.classList.add('is-active')
+
+}
+
+export function closeModal($el) {
+
+  $el.classList.remove('is-active')
+
+}
+
+export function closeAllModals() {
+
+  (document.querySelectorAll('.modal') || []).forEach(($modal) => {
+
+    deleteModal($modal);
+
+  })
+
+}
```
Remove all instances of `$` suggesting JQuery.

Geo-Bold replied:
Done.

anthonyisensee commented:
@Geo-Bold, in progress on my code review for this. You addressed the big issues, so I'm focusing directly on the code this time!

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,172 @@
+}
+
+// Removes a modal from the dom
+export function deleteModal(element, id = undefined) {
+
+  const modalToDelete = id ? document.getElementById(id) : element
+
+  modalToDelete.remove()
+
+}
+
+// Functions to open and close a modal
+export function openModal($el) {
+
+  $el.classList.add('is-active')
+
+}
+
+export function closeModal($el) {
+
+  $el.classList.remove('is-active')
+
+}
+
+export function closeAllModals() {
+
+  (document.querySelectorAll('.modal') || []).forEach(($modal) => {
+
+    deleteModal($modal);
+
+  })
+
+}
```
Using the console, I was able to pull up multiple modals that overlayed each other. That means that an unaware developer could also theoretically create a situation where they'd enable the user to do the same. I don't want this to be possible.

I would suggest making these accessory functions not accept parameters, and set up your code so that it will only ever spawn in one `.modal` (so, if another already exists, newly executing code will kill it or replace the old one's internals). I would consider only ever creating/populating/maintaining one `.modal` with a `#modal` ID. This means that all these functions would not need their `el` or `id` parameters and would make the Modal class far simpler to use in other parts of the code by other developers!

That said, you can also come up with your own way of doing it. But, I only ever want one modal to be able to exist, and I don't want the user to have to worry about closing the old one if they open a new one!

Also, remove "Modal" from function names since external developers will want to call `modal.close()` instead of `modal.closeModal()` in other code files where this code is imported and assigned to `modal`.

I would also recommend that you move these functions to methods inside the Modal class. This will allow you to import the code into Main.js using `import { Modal } from './Modal.js'` and then create new modals using `const modal = new Modal('Do you wish to reset local storage?')` instead of `const confirmReset_1 = new Modal.Modal('Do you wish to reset local storage?')` like you have now.

Geo-Bold replied:
I have included a selection point to prevent a modal from being created while an existing one persists. Instead, any modals are removed from the dom before proceeding to create a new modal.
A private id is now added to the class that is now included in the modal html template string.
Method names have reflected the aforemention changes and all methods are now included inside the class. Import syntax has been changed to reflect this.

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,172 @@
+        <div class="modal is-active">
+            <div class="modal-background"></div>
+            <div class="modal-card">
+                <header class="modal-card-head py-3">
```
Your layout here uses [Bulma spacing helpers](https://bulma.io/documentation/helpers/spacing-helpers/). I'm glad that you found them, but I don't want you to use them. Bulma is a framework for a reason: someone way better than either of us at graphic design looked at all the possible situations (including mobile/desktop/etc layouts) and created default, consistent spacing across the Bulma elements. While we can adjust if we have a _really_ good reason to do so, I think it makes the most sense to stick to defaults.

I didn't like this when I saw it yesterday, but I couldn't quite explain why. Funnily enough, [this typography video](https://www.youtube.com/watch?v=9ElrcTtAxzA&list=LL) came up on my recommended this morning and I decided to watch it. In this video, the narrator shows why this isn't a good idea when it comes to visual weight.

If you would, please remove Bulma spacing helpers across the code unless you have a really good reason to use them.

Geo-Bold replied:
Done! It hurt my eyes too!

anthonyisensee commented on Main.js:
```diff
@@ -165,34 +167,45 @@ document.addEventListener('DOMContentLoaded', () => {
+    document.getElementById('reset_everything').addEventListener("click", () => {
 
-    document.getElementById("reset_everything").addEventListener("click", () => {
+        const confirmReset_1 = new Modal.Modal('Do you wish to reset local storage?')
+        const confirmReset_2 = new Modal.Modal('Resetting will delete all data. Are you sure?')
+        const confirmReset_3 = new Modal.Modal("Please type 'YES' to proceed.", 'Danger', true)
```
I actually didn't know that function (or in this case constructor) parameters could just not be included as desired, and that as long as the function didn't use them (or in this case set defaults) that the function would still run just fine. I learned something new from your code. Thank you!

Geo-Bold replied:
So long as I learn something new everyday, I am content.

anthonyisensee reviewed (changes requested):
Great job on this! Still some work before it's completely ready, but this time around I was able to focus solely on the code instead of broader issues. Please feel free to reply to any individual code review comments if you have questions, and please feel free to challenge me on things I might have called out as not optimal or wrong but that you think are either right or needless critiques.

Geo-Bold committed:
---# bug patch #---
bug: creating a default modal would default to not include a user input field. calling the prompt method would result in no way to get this input.
feat: removes requireUserInput argument when creating a new modal. calling the prompt method now automatically includes a user input field.

anthonyisensee commented:
Included with your next push, add the following after [this line](https://github.com/anthonyisensee/incidenttracker/blob/124eaa1311954d8e428c33b0a375ed8bd86890ba/index.html#L45): `<p>Thank you to the following contributor: <a href="https://github.com/geo-bold">Geo Archbold</a></p>`.

anthonyisensee commented:
@Geo-Bold I changed my mind on where I want your contributor link added. Please do not add another p tag, but instead add `Thank you to <a href="https://github.com/geo-bold">Geo Archbold for contributions</a>.` to the currently existing third p tag [here](https://github.com/anthonyisensee/incidenttracker/blob/10264e973f631e5c437fef0cf2e3c8ee49b1a64f/index.html#L50) after the text currently contained there.

Geo-Bold committed:
Modal Requested Changes Patch
feat: adds author contribution to html footer.
refactor: changes the import to only include the Modal class.
refactor: changes all calls to create a new modal to reflect the import change.
refactor: removes combinations of snake and camel case for variable names.
feat: adds a private modal class id property.
refactor: changes html design to include more build-in bulma functionality for modal styling.
feat: removes modal close events from the header and footer and adds a dedicated close button.
feat: adds an input html string; selection is modified to parse and append the input when the prompt method is called.
refactor: makes use of the nullish co-op instead of guard operators.
feat: adds selection to delete existing modals before a new modal can be created.
feat:adds a private createCloseEvents method to add event listeners to elements that close the modal.
refactor: changes the parameter names for prompt and confirm to be more descriptive.
refactor: changes the selection of the 'OK' buttons to only create event listeners if a function is provided.
feat: adds a close functionality to each 'Cancel' button.
refactor: removes return statements from both prompt and confirm methods.
refactor: renames and moves the deleteModal() into a private method on Modal. it now checks for the unique modal id.
feat: removes depricated or unused functions such as openModal, closeModal and claseAllModals.
style: removes uneccessary whitespace.
feat: adds a modal close button.

Geo-Bold committed:
Modal Requested Changes Patch
feat: adds author contribution to html footer.
refactor: changes the import to only include the Modal class.
refactor: changes all calls to create a new modal to reflect the import change.
refactor: removes combinations of snake and camel case for variable names.
feat: adds a private modal class id property.
refactor: changes html design to include more build-in bulma functionality for modal styling.
feat: removes modal close events from the header and footer.
feat: adds an input html string; selection is modified to parse and append the input when the prompt method is called.
refactor: makes use of the nullish co-op instead of guard operators.
feat: adds selection to delete existing modals before a new modal can be created.
feat:adds a private createCloseEvents method to add event listeners to elements that close the modal.
refactor: changes the parameter names for prompt and confirm to be more descriptive.
refactor: changes the selection of the 'OK' buttons to only create event listeners if a function is provided.
feat: adds a close functionality to each 'Cancel' button.
refactor: removes return statements from both prompt and confirm methods.
refactor: renames and moves the deleteModal() into a private method on Modal. it now checks for the unique modal id.
feat: removes depricated or unused functions such as openModal, closeModal and closeAllModals.
style: removes uneccessary whitespace.

Geo-Bold committed:
feat: adds a modal close button.

Geo-Bold committed:
feat: adds a modal close button.

Geo-Bold commented:
Included in this latest commit are the requested changes. Reach out with any questions or modifications.

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,155 @@
+export class Modal {
+
+    #id = 'modal-version-04/13/2024'
```
I generally don't see dates hard coded into code files. That's what [git blame](https://git-scm.com/docs/git-blame) is for.

anthonyisensee replied:
As one might expect, they can make the code look outdated quickly... I want your code to look timeless!

Geo-Bold replied:
I just wanted to hide a time capsule in the id. Fine, I'll change it to something a little more boring. 😢

anthonyisensee replied:
@Geo-Bold to quote The Hitman's Bodyguard, "Boring is always best!"

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,155 @@
+
+    #id = 'modal-version-04/13/2024'
+
+    modal = `
+        <div class="modal is-active" id="${this.#id}">
+            <div class="modal-background"></div>
+            <button class="modal-close is-large" aria-label="close"></button>
+            <div class="modal-card">
+                <header class="modal-card-head">
+                    <p class="modal-card-title"></p>
+                </header>
+                <section class="modal-card-body">
+                    <div class="field" id="modal-card-field">
+                        <label class="label"></label>
+                    </div>
+                </section>
+                <footer class="modal-card-foot is-justify-content-end">
+                    <div class="buttons">
+                        <button class="button is-danger" id="modal-primary-button"></button>
+                        <button class="button is-info" id="modal-secondary-button"></button>
+                    </div>
+                </footer>
+            </div>
+        </div> 
+    `
```
Thanks for the layout changes! This looks great.

Geo-Bold replied:
You know what? It's growing on me.

anthonyisensee replied:
I CALLED IT!

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,155 @@
+    constructor(message, title, primaryBtnText, secondaryBtnText) {
+
+        this.title = title ?? 'Attention' 
+        this.message = message ?? 'This is the default alert message. Have a great day!'
```
Can we change this default to something that will be semi-meaningful in case the user doesn't set it? For instance, "Are you sure that you want to continue?"

Geo-Bold replied:
I think I'll call this patch: "Serious and Somber: Anthony Edition". 😆

anthonyisensee commented on index.html:
```diff
@@ -42,7 +42,10 @@ <h2>Danger Zone</h2>
-            <p>Developed by <a href="https://github.com/anthonyisensee">Anthony Isensee</a> for Ricky Moore. 🖤</p>
+            <p>
+                Developed by <a href="https://github.com/anthonyisensee">Anthony Isensee</a> for Ricky Moore. 🖤
+                Thank you to <a href="https://github.com/geo-bold">Geo Archbold</a> for contributions.
```
Truly, thank you!! I've had a ton of fun with this.

Geo-Bold replied:
It has been an absolute pleasure! Thank you!

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,155 @@
+
+    #createTemplate() {
+        
+        const existingModal = document.getElementById(`${this.#id}`)
```
Really like how you're referencing one record of the ID!

Geo-Bold replied:
I liked your idea. Makes garbage collection a little more private and less prone to accidents.

anthonyisensee replied:
Absolutely!

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,155 @@
+
+    }
+
+    prompt(funcToRunOnOkay, funcToRunOnCancel) {
```
Renaming these parameters increases readability _so much!_

Geo-Bold replied:
You were right! I have to overcome my fear of choosing names that are more than ten characters long. Descriptions make the code way more serviceable.

anthonyisensee replied:
So true!

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,155 @@
+
+        }
+
+        secondaryButton.addEventListener('click', () => {
+
+            if (funcToRunOnCancel) {
+
+                funcToRunOnCancel(userInput.value)
+                
+            } else { this.#close() }
+
+        })
```
I think that the cancel button should always close the modal _and_ run the function if it exists. What do you think?

Geo-Bold replied:
Good point, making the change now.

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,155 @@
+
+        const modal = document.getElementById(this.#id)
+
+        modal && modal.remove()
```
Do you think the one liner `if (modal) modal.remove()` is more readable than using a logical operator combined with short circuiting logical evaluation?

Geo-Bold replied:
I'm afraid I can't respond since I have died on this hill.

anthonyisensee replied:
😂😂

anthonyisensee reviewed (commented):
Great job!! This code has improved 10x from your initial commit. I left several comments, but none of them are truly necessary. Let me know once you've reviewed, made any changes (not required), and are ready for me to merge this into prod!

Geo-Bold committed:
Somber and Serious Patch
refactor: alters the modal id to 'new.modal'.
refactor: changes the default message to something more ubiquitous.
feat: adds the close method to the ok and cancel buttons.

Geo-Bold committed:
Merge branch 'prod' into bulma-modal-prompts

anthonyisensee commented on Modal.js:
```diff
@@ -0,0 +1,158 @@
+export class Modal {
+
+    #id = 'new.modal'
```
`document.querySelector("#new.modal")`:

![giphy](https://github.com/anthonyisensee/incidenttracker/assets/60198052/7076a14d-40dc-425b-868b-af8df074cdbf)

See the following run from the console:

![image](https://github.com/anthonyisensee/incidenttracker/assets/60198052/3b51a226-a7b0-4c6e-8a2b-852434e775eb)

Using periods in IDs is allowed, but [it will cause you additional pain in the long run](https://stackoverflow.com/a/79022/13173288). They are so rarely used that I literally had to double check to make sure they are allowed as I've never seen them in IDs or classes before. I suggest changing to simply `modal`. There's only every supposed to be one of them, after all!

I think this is the final change!!

Geo-Bold replied:
Change has been made.

anthonyisensee reviewed (changes requested):
You're so, _so_, **so** close. Let me know when the final change has been made.

Geo-Bold committed:
refactor: changes modal to 'modal'. removes period separator.

anthonyisensee commented:
Whoop whoop! It's live, baby! @Geo-Bold, I imagine you must feel something like this:

![bowl-girl](https://github.com/anthonyisensee/incidenttracker/assets/60198052/70b861b9-c232-498d-925c-593f63977220)

Geo-Bold commented:
![More like...](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExMTBrYTJyZ3llYnRlemZ3bXVvOW0wNzdqcHV5bndqdDZidTBjaHI3NyZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/MOWPkhRAUbR7i/giphy.gif)
So excited to be done!
